Adds support for building Wasm on Windows

steveisok opened the pull request:
Emscripten and EMSDK_PATH must be set up externally.  This differs from linux and macos where we optionally can do it for you.

Commit:
First pass at building wasm on windows

Commit:
Adds support for building wasm on Windows

Emscripten and EMSDK_PATH must be set up externally.  This differs from linux and macos
where we optionally can do it for you.

steveisok commented:
@lewing @akoeplinger Thoughts on where this should build on CI?

jkotas commented on eng/native/gen-buildsys.cmd:
```diff
@@ -35,7 +35,7 @@ if /i "%__Ninja%" == "1" (
         if /i "%__Arch%" == "x86" (set __ExtraCmakeParams=%__ExtraCmakeParams% -A Win32)
     ) else (
-        set __CmakeGenerator=NMake Makefiles
+        set __CmakeGenerator=Ninja
```
Accidental change?

steveisok replied:
Indeed it was

jkotas commented:
cc @yowl

jkotas commented on src/libraries/Native/build-native.cmd:
```diff
@@ -97,7 +97,12 @@ goto :SetupDirs
+
+if /i "%__BuildArch%" == "wasm" (
+    set __sourceDir=%~dp0\Unix
+    set __ExtraCmakeParams=%__ExtraCmakeParams% "-DCLR_ENG_NATIVE_DIR=%__engNativeDir%" "-DCMAKE_BUILD_TYPE=%CMAKE_BUILD_TYPE%"
```
`CLR_ENG_NATIVE_DIR` comes from `eng\native\configurepaths.cmake` in other places. Can this be on the same plan?

steveisok replied:
I think this was hanging around previously and thought it was necessary.  I saw that configurepaths.cmake was already included, so taking it out worked.

Commit:
Cleanup

akoeplinger commented on src/mono/mono.proj:
```diff
@@ -477,12 +478,18 @@
         <Destination>$(RuntimeBinDir)cross\llc</Destination>
       </_MonoRuntimeArtifacts>
-      <_MonoRuntimeArtifacts Condition="'$(MonoAOTBundleLLVMOptimizer)' == 'true'" Include="$(MonoAOTLLVMDir)\bin\opt">
+      <_MonoRuntimeArtifacts Condition="'$(MonoAOTBundleLLVMOptimizer)' == 'true' and '$(OS)' == 'Windows_NT'" Include="$(MonoAOTLLVMDir)\bin\llc.exe">
+        <Destination>$(RuntimeBinDir)cross\llc.exe</Destination>
```
an easier approach I've seen elsewhere is something like this:
```xml
<_MonoRuntimeArtifacts Condition="'$(MonoAOTBundleLLVMOptimizer)' == 'true'" Include="$(MonoAOTLLVMDir)\bin\llc$(ExeExtension)">
  <Destination>$(RuntimeBinDir)cross\llc$(ExeExtension)</Destination>
```

and then setting ExeExtension to .exe only on Windows.

akoeplinger commented on src/mono/wasm/wasm.proj:
```diff
@@ -8,6 +8,7 @@
     <!-- FIXME: clean up the duplication with libraries Directory.Build.props -->
     <PackageRID>browser-wasm</PackageRID>
     <NativeBinDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'native', '$(NetCoreAppCurrent)-$(TargetOS)-$(Configuration)-$(TargetArchitecture)'))</NativeBinDir>
+    <ICULibDir>$(PkgMicrosoft_NETCore_Runtime_ICU_Transport)/runtimes/browser-wasm/native/lib</ICULibDir>
```
please follow the patter of the other props and use `$([MSBuild]::NormalizeDirectory`

akoeplinger commented on src/mono/wasm/wasm.proj:
```diff
@@ -104,4 +102,109 @@
+  <UsingTask TaskName="GenerateWasmVersionFile"
+    AssemblyFile="$(WasmBuildTasksAssemblyPath)"/>
+  <Target Name="BuildWasmRuntimes"
+          Condition="'$(OS)' == 'Windows_NT'"
```
this should be used everywhere, not just Windows :)

vargaz replied:
If we really want to build this on windows for some reason, we should move to a cmake based build instead of this.

steveisok replied:
Let's get this in as a first step and then we can switch later.

Commit:
Feedback

Commit:
Rename conflicting target

lewing commented:
/ccc @radical

jkotas commented:
> Thoughts on where this should build on CI?

We would need to get a Windows machine with Emscripten. The easiest way to do that is to install Emscripten on the fly. https://github.com/dotnet/runtimelab/blob/d0cda961aeda26faae17c474e26d822cd1e45438/eng/pipelines/runtimelab/install-emscripten.cmd is a small script to do that. I am not sure whether it would be appropriate for dotnet/runtime CI though.

akoeplinger commented:
> Thoughts on where this should build on CI?

I think we'll mostly want to protect the Windows wasm build for local developer purposes, official builds will likely always use Linux so to me this would make it a good fit for the `runtime-live-build` pipeline.

steveisok commented:
> > Thoughts on where this should build on CI?
> 
> I think we'll mostly want to protect the Windows wasm build for local developer purposes, official builds will likely always use Linux so to me this would make it a good fit for the `runtime-live-build` pipeline.
> 
> I'm not a big fan of introducing the on the fly download of Emscripten like you do in the runtimelab since it's another dependency that can break dotnet/runtime CI.

Agreed re: not downloading emscripten every time.  

@lewing @radical Do you see the need to run tests?  Or will only build validation suffice?

Commit:
Merge remote-tracking branch 'upstream/master' into wasm-windows-build

Commit:
Made sure current master still works

Commit:
auto provision when EMSDK_PATH is not set and try to see if a browser build works on windows

Commit:
Add call in front of batch scripts. All of a sudden it's necessary

Commit:
Merge remote-tracking branch 'upstream/main' into wasm-windows-build

radical commented on src/mono/wasm/wasm.proj:
```diff
@@ -110,4 +108,109 @@
 
+  <UsingTask TaskName="GenerateWasmVersionFile"
+    AssemblyFile="$(WasmBuildTasksAssemblyPath)"/>
+  <Target Name="BuildWinWasmRuntimes"
```
Can this target replace the `make` invocation in `BuildWasmRuntimes`? I mean, are they two effectively the same?

steveisok replied:
Yes

radical commented on src/mono/wasm/wasm.proj:
```diff
@@ -110,4 +108,109 @@
+      <ICULibFiles Include="$(ICULibDir)/*.dat" />
+    </ItemGroup>
+
+    <Exec Command="$(EmccCmd) $(EmccFlags) $(EmccConfigurationFlags) -Oz -I$(MonoIncludeDir) runtime/corebindings.c -c -o $(MonoObjDir)corebindings.o"
```
The `$(EmccCmd)` invocations can now use `RunWithEmSdk` task.

radical commented on src/mono/wasm/wasm.proj:
```diff
@@ -110,4 +108,109 @@
+    <Exec Command="$(EmccCmd) --version > $(WasmVersionFile)"
+          IgnoreStandardErrorWarningFormat="true" />
+    
+    <GenerateWasmVersionFile SourceFile="$(WasmVersionFile)"
```
This can be done like in `WasmApp.targets`:

```xml
    <RunWithEmSdkEnv Command="emcc --version" WorkingDirectory="$(_WasmIntermediateOutputPath)" EmSdkPath="$(EMSDK_PATH)" ConsoleToMsBuild="true">
      <Output TaskParameter="ConsoleOutput" ItemName="_VersionLines" />
    </RunWithEmSdkEnv>
```

radical commented on src/mono/mono.proj:
```diff
@@ -89,6 +90,27 @@
+      <EMSDK_PATH>$(ProvisionEmscriptenDir)</EMSDK_PATH>
+      <WasmLocalPath>$([MSBuild]::NormalizeDirectory('$(MSBuildThisFileDirectory)', 'wasm'))</WasmLocalPath>
+      <EmsdkLocalPath>emsdk</EmsdkLocalPath>
+      <EmscriptenVersion>2.0.11</EmscriptenVersion>
```
We should move this to a file, so it can be read from the Makefile, and here.

radekdoulik committed:
Use RunWithEmSdkEnv task

radekdoulik committed:
Override arch for Browser OS

Similar to what we do in build.sh

radekdoulik committed:
Merge remote-tracking branch 'upstream/main' into wasm-windows-build

radekdoulik committed:
Use msbuild to write the emcc version file

radekdoulik committed:
Update machine certificates

To avoid this error:

    Installing tool 'node-12.18.1-64bit'..
      Installation failed!
      Error: Downloading URL 'https://storage.googleapis.com/webassembly/emscripten-releases-builds/deps/node-v12.18.1-win-x64.zip': <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1045)>
      Warning: Possibly SSL/TLS issue. Update or install Python SSL root certificates (2048-bit or greater) supplied in Python folder or https://pypi.org/project/certifi/ and try again.
      Warning: The SDK/tool 'node-12.18.1-64bit' cannot be activated since it is not installed! Skipping this tool...

The script comes from runtimelabs repo https://github.com/dotnet/runtimelab/blob/d0cda961aeda26faae17c474e26d822cd1e45438/eng/pipelines/runtimelab/update-machine-certs.ps1

radekdoulik committed:
Put emscripten version in text file

And use it in Makefile and mono.proj

radical commented:
> @lewing @radical Do you see the need to run tests? Or will only build validation suffice?

IMHO, for this PR, build validation should be enough. That should enable more of us to try this out, and tests can be enabled in a follow up PR.

radekdoulik committed:
Try to fix Browser_wasm_Windows detection

radekdoulik committed:
Install native dependencies

I think we are missing cmake and failing with:

      configure: cmake -DENABLE_WERROR=1 -DCMAKE_INSTALL_PREFIX=D:\workspace\_work\1\s\artifacts\obj\mono\Browser.wasm.Debug\out -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Debug -DGC_SUSPEND=preemptive -DENABLE_MINIMAL=jit,portability,sgen_major_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,sgen_debug_helpers,sgen_binary_protocol,logging,shared_perfcounters,interpreter,threads,eventpipe,qcalls -DENABLE_INTERP_LIB=1 -DDISABLE_ICALL_TABLES=1 -DDISABLE_CRASH_REPORTING=1 -DENABLE_ICALL_EXPORT=1 -DENABLE_LAZY_GC_THREAD_CREATION=1 -DENABLE_LLVM_RUNTIME=1 -DENABLE_METADATA_UPDATE=1 "-DCMAKE_C_FLAGS= -fexceptions -ID:\workspace\_work\1\s\.packages\microsoft.netcore.runtime.icu.transport\6.0.0-preview.3.21151.1/runtimes/browser-wasm/native/include" "-DCMAKE_CXX_FLAGS= -fexceptions" D:\workspace\_work\1\s\src\mono\ -DCMAKE_TOOLCHAIN_FILE=D:\workspace\_work\1\s\src\mono\wasm\emsdk\upstream\emscripten\cmake\Modules\Platform\Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR="D:/workspace/_work/1/s/src/mono/wasm/emsdk/node/12.18.1_64bit/bin/node.exe" -G Ninja
      Traceback (most recent call last):
        File "D:\workspace\_work\1\s\src\mono\wasm\emsdk\upstream\emscripten\\emcmake.py", line 33, in <module>
          sys.exit(run())
        File "D:\workspace\_work\1\s\src\mono\wasm\emsdk\upstream\emscripten\\emcmake.py", line 26, in run
          building.configure(sys.argv[1:])
        File "D:\workspace\_work\1\s\src\mono\wasm\emsdk\upstream\emscripten\tools\building.py", line 339, in configure
          run_process(args, stdout=stdout, stderr=stderr, env=env, **kwargs)
        File "D:\workspace\_work\1\s\src\mono\wasm\emsdk\upstream\emscripten\tools\shared.py", line 94, in run_process
          ret = subprocess.run(cmd, check=check, input=input, *args, **kw)
        File "D:\obj\windows-release\37amd64_Release\msi_python\zip_amd64\subprocess.py", line 472, in run
        File "D:\obj\windows-release\37amd64_Release\msi_python\zip_amd64\subprocess.py", line 775, in __init__
        File "D:\obj\windows-release\37amd64_Release\msi_python\zip_amd64\subprocess.py", line 1178, in _execute_child
      FileNotFoundError: [WinError 2] The system cannot find the file specified
    D:\workspace\_work\1\s\src\mono\mono.proj(410,5): error MSB3073: The command "call D:\workspace\_work\1\s\src\mono\wasm\emsdk\emsdk_env.bat && emcmake cmake -DENABLE_WERROR=1 -DCMAKE_INSTALL_PREFIX=D:\workspace\_work\1\s\artifacts\obj\mono\Browser.wasm.Debug\out -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Debug -DGC_SUSPEND=preemptive -DENABLE_MINIMAL=jit,portability,sgen_major_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,sgen_debug_helpers,sgen_binary_protocol,logging,shared_perfcounters,interpreter,threads,eventpipe,qcalls -DENABLE_INTERP_LIB=1 -DDISABLE_ICALL_TABLES=1 -DDISABLE_CRASH_REPORTING=1 -DENABLE_ICALL_EXPORT=1 -DENABLE_LAZY_GC_THREAD_CREATION=1 -DENABLE_LLVM_RUNTIME=1 -DENABLE_METADATA_UPDATE=1 -DCMAKE_C_FLAGS=" -fexceptions -ID:\workspace\_work\1\s\.packages\microsoft.netcore.runtime.icu.transport\6.0.0-preview.3.21151.1/runtimes/browser-wasm/native/include" -DCMAKE_CXX_FLAGS=" -fexceptions"  D:\workspace\_work\1\s\src\mono\" exited with code 1.
    ##[error]src\mono\mono.proj(410,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command "call D:\workspace\_work\1\s\src\mono\wasm\emsdk\emsdk_env.bat && emcmake cmake -DENABLE_WERROR=1 -DCMAKE_INSTALL_PREFIX=D:\workspace\_work\1\s\artifacts\obj\mono\Browser.wasm.Debug\out -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Debug -DGC_SUSPEND=preemptive -DENABLE_MINIMAL=jit,portability,sgen_major_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,sgen_debug_helpers,sgen_binary_protocol,logging,shared_perfcounters,interpreter,threads,eventpipe,qcalls -DENABLE_INTERP_LIB=1 -DDISABLE_ICALL_TABLES=1 -DDISABLE_CRASH_REPORTING=1 -DENABLE_ICALL_EXPORT=1 -DENABLE_LAZY_GC_THREAD_CREATION=1 -DENABLE_LLVM_RUNTIME=1 -DENABLE_METADATA_UPDATE=1 -DCMAKE_C_FLAGS=" -fexceptions -ID:\workspace\_work\1\s\.packages\microsoft.netcore.runtime.icu.transport\6.0.0-preview.3.21151.1/runtimes/browser-wasm/native/include" -DCMAKE_CXX_FLAGS=" -fexceptions"  D:\workspace\_work\1\s\src\mono\" exited with code 1.

radekdoulik committed:
Try use newer cmake

radekdoulik committed:
Revert "Try use newer cmake"

This reverts commit bd95f346f9a3c5828924201c79df3b356043c11c.

ivan-prodanov commented:
Can't wait to use this. 👀

radekdoulik committed:
Normalize path

radekdoulik committed:
Remove excessive space char

radekdoulik committed:
Try newer emscripten version

To try if it helps with the cmake issue

radekdoulik committed:
Try trim end '\' from directory passed to cmake

Also revert emscripten version back

radekdoulik committed:
Provision emscripten only when targeting wasm

radekdoulik committed:
Do not set $os, it is already set above

radekdoulik committed:
Revert "Remove excessive space char"

This reverts commit b1c4da12a0c39e78b0cf02d1da1332a55cedf772.

radekdoulik committed:
Provision emscripten only on windows

radical reviewed (approved):
Looks good. There are some issues like it breaks if the runtime checkout is in a path with spaces, but we can fix those in a follow up PR. CI doesn't have the same issue.

ivan-prodanov commented:
Great job everyone!
Do we have an example on how to build Wasm apps on windows?

radekdoulik commented on src/mono/mono.proj:
```diff
@@ -360,11 +386,12 @@
 
     <PropertyGroup>
-      <_MonoCMakeConfigureCommand>cmake @(_MonoCMakeArgs, ' ') $(MonoCMakeExtraArgs) $(MonoProjectRoot)</_MonoCMakeConfigureCommand>
+      <_MonoCMakeConfigureCommand>cmake @(_MonoCMakeArgs, ' ') $(MonoCMakeExtraArgs) $(MonoProjectRoot.TrimEnd('\/'))</_MonoCMakeConfigureCommand>
```
I didn't find any. There are functions that do the opposite - make sure that there is trailing slash. @radical do you know if there's better way to do this?

radekdoulik commented on src/mono/mono.proj:
```diff
@@ -18,6 +18,7 @@
     <ExeExt Condition="'$(OS)' == 'Windows_NT'">.exe</ExeExt>
     <ScriptExt Condition="'$(OS)' == 'Windows_NT'">.cmd</ScriptExt>
     <ScriptExt Condition="'$(OS)' != 'Windows_NT'">.sh</ScriptExt>
+    <ExeExt Condition="'$(OS)' == 'Windows_NT'">.exe</ExeExt>
```
indeed. fixed in https://github.com/dotnet/runtime/pull/49771

radekdoulik commented on src/mono/wasm/Makefile:
```diff
@@ -33,7 +33,7 @@ all: build-native icu-files source-files header-files
 #
 
-EMSCRIPTEN_VERSION=2.0.12
+EMSCRIPTEN_VERSION=`cat emscripten-version.txt`
```
fixed in https://github.com/dotnet/runtime/pull/49771

radekdoulik commented on eng/pipelines/runtime-staging.yml:
```diff
@@ -217,3 +217,14 @@ jobs:
+    jobParameters:
+      testGroup: innerloop
+      nameSuffix: Browser_wasm_Windows
+      buildArgs: -subset mono+libs /p:RuntimeFlavor=Mono
```
yes, looking forward to use the nuget instead
